src/communication/websocket/session.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "communication/websocket/session.hpp"
+
+#include "utils/logging.hpp"
+
+namespace communication::websocket {
+namespace {
+void LogError(boost::beast::error_code ec, const std::string_view what) {
+  spdlog::warn("Websocket session failed on {}: {}", what, ec.message());
+}
+}  // namespace
+
+void Session::Run() {
+  ws_.set_option(boost::beast::websocket::stream_base::timeout::suggested(boost::beast::role_type::server));
+
+  ws_.set_option(boost::beast::websocket::stream_base::decorator(
+      [](boost::beast::websocket::response_type &res) { res.set(boost::beast::http::field::server, "Memgraph WS"); }));
+
+  // Accept the websocket handshake
+  boost::beast::error_code ec;
+  ws_.accept(ec);
+  if (ec) {
+    return LogError(ec, "accept");
+  }
+  connected_.store(true, std::memory_order_relaxed);
+
+  // run on the strand
+  boost::asio::dispatch(strand_, [shared_this = shared_from_this()] { shared_this->DoRead(); });
+}
+
+void Session::Write(std::shared_ptr<std::string> message) {
+  if (!connected_.load(std::memory_order_relaxed)) {
+    return;
+  }
+  boost::asio::dispatch(strand_, [message, shared_this = shared_from_this()] {
+    shared_this->messages_.push_back(message);
+
+    if (shared_this->messages_.size() > 1) {
+      return;
+    }
+    shared_this->DoWrite();
+  });
+}
+
+bool Session::IsConnected() const { return connected_.load(std::memory_order_relaxed); }
+
+void Session::DoWrite() {
+  const auto next_message = messages_.front();
+  ws_.async_write(boost::asio::buffer(*next_message), [message_string = next_message, shared_this = shared_from_this()](
+                                                          boost::beast::error_code ec, const size_t bytes_transferred) {
+    shared_this->OnWrite(ec, bytes_transferred);
```

Comment:
This won't execute on `strand_`, so I think this might cause some data race between line 66 and line 46.

Reply:
why do you think it won't execute on `strand_`? It's DoWrite is dispatched on it.

Reply:
Oh I see, it doesn't fetch automatically the executor of the write, I should bind the executor of the callback to strand_.

Reply:
It seems I was right the first time:
https://stackoverflow.com/a/15879971
It will execute everything on the strand. YOu also get a compiler error if you try to bind strand to the completion handler explicitly.